### PR TITLE
Changes to fields and description of NXdisk_chopper

### PR DESCRIPTION
**Pull request by matthew-d-jones**
~Since the existing `NXdisk_chopper` definitions appeared not to be fit for purpose I have been bold in making breaking changes.~ edit: changes are now non-breaking

Changes follow discussion of requirements for recording chopper data at the ESS.

I think the `delay` description requires further refinement.

### CODE REVIEW

**Comment by matthew-d-jones** [comment]
@zjttoefs I've created this as a draft PR for you to take a look at first. The `phase` description still makes little sense to me.

**Comment by zjttoefs** [comment]
Is there a need to make the changes breaking?

I don't think we need to remove anything. And the phase in time might better be called _delay_. No?

File: base_classes/NXdisk_chopper.nxdl.xml [hunk 1]
```diff
@@ -24,56 +24,78 @@
...
+		<dimensions><dim index="1" value="2n"/></dimensions>
 	</field>
-	<field name="pair_separation" type="NX_FLOAT" units="NX_LENGTH">
-		<doc>disc spacing in direction of beam</doc>
+	<field name="top_dead_center" type="NX_NUMBER" units="NX_TIME">
+		<doc>
+			Timestamps of the top-dead-center signal
+		</doc>
+	</field>
+	<field name="beam_position" type="NX_FLOAT" units="NX_ANGLE">
```
Comment (zjttoefs) [inline base_classes/NXdisk_chopper.nxdl.xml]: I don't think we need this if we record the slit pattern starting from the TDC angle.

**Comment by matthew-d-jones** [comment]
The other breaking changes were removing the `type` and `wavelength_range` fields which don't make sense for a single disk, and changing the meaning of the radius field which made little sense to me.

File: base_classes/NXdisk_chopper.nxdl.xml (hunk 1 above)
Comment (matthew-d-jones) [reply base_classes/NXdisk_chopper.nxdl.xml]: That depends whether the TDC position means the "position of the TDC sensor" or the "position of the beam when the disk is in the position where the TDC timestamp is recorded".

Comment (matthew-d-jones) [reply base_classes/NXdisk_chopper.nxdl.xml]: Maybe the latter is more intuitive, just clunkier to describe.

**Comment by matthew-d-jones** [comment]
Just realised I carefully used American spelling everywhere to be consistent with "disk" in "NXdisk_chopper" but then used the term "anticlockwise".

File: base_classes/NXdisk_chopper.nxdl.xml (hunk 1 above)
Comment (prjemian) [reply base_classes/NXdisk_chopper.nxdl.xml]: +1 it's an international standard, TDC is jargon (known in automotive)

**Comment by prjemian** [comment]
Re [wavelength_range](https://github.com/nexusformat/definitions/pull/649#issuecomment-474414023), isn't there always a range of wavelengths that depends on the chopper opening size and the rotation speed?

**Comment by matthew-d-jones** [comment]
> Re [wavelength_range](https://github.com/nexusformat/definitions/pull/649#issuecomment-474414023), isn't there always a range of wavelengths that depends on the chopper opening size and the rotation speed?

I suppose there is technically a low bound on the energy due to the thickness of the disk, but otherwise not really with a single disk.

**Comment by mkoennecke** [comment]
There is a contradiction here: in the top documentation it says that there ought to be a NXdisk_chopper group per disk. Then the pair_separation does not make any sense as it implies a chopper pair. 

Regarding the phase. Given there is a distance between two choppers. Then the second chopper has to open a little later then the first one in order to account for the flight time between the two choppers. This is expressed as an angular offset, the phase, in the chopper systems I had to deal with so far. This is something  we read from the chopper controller, thus I want to have this field in order to record it. I do not mind the additional fields, they serve to describe the chopper in a more accurate way.

**Comment by matthew-d-jones** [comment]
@mkoennecke 
There is already an angular `phase` field.

I originally removed the fields which don't make sense for a single disk, but Tobias is keen not to make breaking changes at the same time.
Given that there is nowhere else to store potentially useful information like `wavelength_range`, perhaps it is not unreasonable to store it in one, or both, of the two separate `NXdisk_chopper`s making up the pair? Similarly, specifying `type` of "synchro_pair" could imply that this disk is one of those in the pair. We would need to change the wording to make this clear.

**Comment by prjemian** [comment]
Ready to merge?

**Comment by zjttoefs** [comment]
I have given my approval. If no one else comments with requests for changes in the next days, I would merge.

**Comment by prjemian** [comment]
Ok, then.  If no additional comments by May 22, then I will merge this.

**Comment by mkoennecke** [comment]
I am mostly OK with this. But I have a minor point: There is this slits field which indicates that a chopper may have more than one slit (They usually do). Should then slit_angle and slit_edges not be arrays of length nSlits? This might apply to the radius etc. fields too.

**Comment by matthew-d-jones** [comment]
> I am mostly OK with this. But I have a minor point: There is this slits field which indicates that a chopper may have more than one slit (They usually do). Should then slit_angle and slit_edges not be arrays of length nSlits? This might apply to the radius etc. fields too.

slit_edges is already described as an array of length 2*nSlits:
https://github.com/nexusformat/definitions/pull/649/files#diff-965d206bd4d315d80b5f08de5a97d475R79

but I forgot to add the array description for `slit_angle` and `slit_height`, thanks for catching that.

**Comment by zjttoefs** [comment]
I don't like the new change. It changes existing fields and hence breaks existing files.

The height is rarely different for different cut outs, I believe. And changing the slit angle to an array makes the slit_edges redundant.

**Comment by prjemian** [comment]
Seems this is not yet ready to merge.

File: base_classes/NXdisk_chopper.nxdl.xml [hunk 2]
```diff
@@ -24,13 +24,26 @@
...
+	<symbols>
+    <doc>This symbol will be used below to coordinate datasets with the same shape.</doc>
+    <symbol name="n"><doc>Number of slits in the disk</doc></symbol>
+  </symbols>
+
 	<doc>
-		A device blocking the beam in a temporal periodic pattern.
-		
-		TODO: need documentation
+		A disk which blocks the beam but has one or more slits to periodically
```
Comment (prjemian) [inline base_classes/NXdisk_chopper.nxdl.xml]: First line of this documentation should be a single-line summary (to be used in creating this summary: http://download.nexusformat.org/doc/html/classes/base_classes/index.html#base-class-definitions).  A blank line afterwards, then any multi-line text.

**Comment by zjttoefs** [comment]
Since @mkoennecke  framed his suggestions yesterday as "minor comments" I would think we're okay to merge after reverting the last commit and addressing the documentation improvement that @prjemian has highlighted.

File: base_classes/NXdisk_chopper.nxdl.xml [hunk 3]
```diff
@@ -42,37 +57,65 @@
 		</enumeration>
 	</field>
 	<field name="rotation_speed" type="NX_FLOAT" units="NX_FREQUENCY">
-		<doc>chopper rotation speed</doc>
+		<doc>
+			Chopper rotation speed. Positive for anticlockwise rotation when
+			facing the sample, negative otherwise.
```
Comment (zjttoefs) [inline base_classes/NXdisk_chopper.nxdl.xml]: I would make this "facing away from the source". Removes confusion if you ever put a chopper after the sample.

File: base_classes/NXdisk_chopper.nxdl.xml [hunk 4]
```diff
@@ -42,37 +57,65 @@
...
+	<field name="slit_edges" type="NX_FLOAT" units="NX_ANGLE">
+		<doc>
+			Angle of the each edge of each slit from the position of the
+			top-dead-center timestamp sensor, anticlockwise when facing the sample.
+			The first edge must be the opening edge of a slit, thus the last edge
+			may have an angle greater than 360 degrees.
+		</doc>
+		<dimensions><dim index="1" value="2n"/></dimensions>
+	</field>
+	<field name="top_dead_center" type="NX_NUMBER" units="NX_TIME">
```
Comment (zjttoefs) [inline base_classes/NXdisk_chopper.nxdl.xml]: We should allow a start_time attribute (or whatever is consistent with NXlog).

File: base_classes/NXdisk_chopper.nxdl.xml (hunk 3 above)
Comment (matthew-d-jones) [reply base_classes/NXdisk_chopper.nxdl.xml]: :+1: done

File: base_classes/NXdisk_chopper.nxdl.xml (hunk 4 above)
Comment (matthew-d-jones) [reply base_classes/NXdisk_chopper.nxdl.xml]: :+1: done

**Comment by zjttoefs** [comment]
Since we have three green ticks from reviewers, I would almost feel comfortable to merge. Especially since there are no changes (let alone breaking ones) to existing definitions.
But I'll give it a day or so, if @prjemian or @mkoennecke want to reassess the PR.

File: base_classes/NXdisk_chopper.nxdl.xml [hunk 5]
```diff
@@ -24,56 +24,76 @@
...
+			facing the sample, negative otherwise.
+		</doc>
 	</field>
 	<field name="slits" type="NX_INT">
 		<doc>Number of slits</doc>
 	</field>
 	<field name="slit_angle" type="NX_FLOAT" units="NX_ANGLE">
-		<doc>angular opening</doc>
+		<doc>
+			Angle of the each edge of each slit from the position of the
```
Comment (prjemian) [inline base_classes/NXdisk_chopper.nxdl.xml]: change `the each` to `each` and I'm good overall

Comment (zjttoefs) [reply base_classes/NXdisk_chopper.nxdl.xml]: To cut this short, while I can't edit @matthew-d-jones source for the PR, I'll merge and drop the "the" in master.

**Comment by zjttoefs** [comment]
Brilliantly referenced the wrong pull request, but the deed is done in 935afe4bd3ee0bdd458cd9cfa61bed731ae62723

**Pull request merged**